### ISSUE TITLE
support proxy authentication

### ISSUE THREAD

**Issue by techieshark** [mapbox/mapbox-studio-classic#1315]
I'm working behind a proxy which requires authentication and Mapbox Studio isn't working in that context.

(I'd love to just change the network settings but I'm working on a project for the Dept of Education in New South Wales, Australia and don't have immediate control over the network.)

In issue #860 (general proxy support), there is [a comment](https://github.com/mapbox/mapbox-studio/issues/860#issuecomment-65721792) saying "no proxy authentication yet." Since it seems this may be a sufficiently different case than the bare minimum of getting Mapbox Studio working through a non-authentication-required proxy, I'll make a different issue for it here, but feel free to close this if you think it is too close to being a duplicate.

Thanks!

cc: @Rustuma @Lilith-Palmer

**Comment by wilhelmberg** [comment]
@techieshark can you share some details about the proxy?
Do you know if it's squid or another software?
Are there PAC scripts being used?
How do you authenticate with other programs?

**Comment by techieshark** [comment]
@BergWerkGIS according to whatismyproxy.com, "The header: HTTP_X_BLUECOAT_VIA is present", which seems to indicate we're behind a Blue Coat system. We don't use any proxy auto config / PAC scripts. I'm on a Mac, and have the proxy address + my authentication info in the system's Network Preferences.

Chrome uses the system's proxy config automatically. Command line programs I use like git have their own proxy configuration settings in their config files.

**Comment by wilhelmberg** [comment]
@techieshark thanks for providing additional information.

I'm planning on looking into the proxy issue again as soon as I've set up one for testing (hopefully next week).

Although I'm not confident that Blue Coat will be supported any time soon.
Details about Blue Coat were already discussed in #860.

Short version: it seems to have an internal white list of user agents.
If a software doesn't have a proper user agent set, that software gets blocked.
Not all node modules we use, set proper user agents.
Once I've identified all modules that don't set proper user agents I will contact the maintainers and try to get the changes upstream.
Might take some time.

**Comment by techieshark** [comment]
@BergWerkGIS Interesting! Thank you for the info and for looking into this. Let me know should you need anything else from my end.

**Comment by ghost** [comment]
Bluecoat NTLM here. Mapbox Studio is logging a 500 error. Chrome/Chromium should have no issues authenticating with our proxy (I'm posting from a Chrome session now).

```
HTTP/1.1 500 Internal Server Error
X-Powered-By: Express
Content-Type: text/html; charset=utf-8
Content-Length: 2003
ETag: W/"LjYer5Nz3hHSeoNg19ovdw=="
Date: Tue, 12 May 2015 01:52:44 GMT
Connection: keep-alive
```

And in the log:

```
<Tue, 12 May 2015 01:43:53 GMT> error 500 { [Error: Cannot find Mapbox API at https://a.tiles.mapbox.com]
  method: 'GET',
  url: '/oauth/config?MapboxAPITile=https%3A%2F%2Fa.tiles.mapbox.com' }
```

Notes:
Users here _do not_ know the proxy address, username or password. The proxy address is done via DHCP/PAC/Group Policy and credentials are grabbed using SSPI. Again, Chrome/Chromium should have no issues with this as it queries the Windows proxy config.

**Comment by wilhelmberg** [comment]
@voltagex sorry I missed your post, could you try to just start Mabox Studio (don't do anything there) and then go to Chrome http://localhost:3000 ?
Any success?

**Comment by wilhelmberg** [comment]
@voltagex 
Would be great if you could try this proxy-enabled version of Mapbox Studio linked here:
https://github.com/mapbox/mapbox-studio/issues/860#issuecomment-65508867
and tell me, if it works.

But beware, it is quite old (Dec 2014).

At the moment, I don't see any chance to support Bluecoat in the near future as one of its features seems to be blocking programs with uncommon user agents.
Mapbox Studio relies on a lot of node modules that are out of our control and don't provide any or invalid (as seen by Bluecoat) user agent strings.

**Comment by ghost** [comment]
Will try it tomorrow. It's not even connecting to the proxy at this point
but I don't believe user agent filtering is in effect where I am.

**Comment by ghost** [comment]
@BergWerkGIS this "works" - as in it hits my local chaining proxy (which is handling NTLM auth) and gets to the OAuth screen. Is it possible to provide a newer build? It goes back to the first Connect screen after I hit "Allow".

**Comment by techieshark** [comment]
@BergWerkGIS bummer, but thanks for looking into this. 

Bluecoat's strategy seems better at preventing people from using legitimate (but non standard) tools like Mapbox Studio rather than enforcing real security - it sounds like if a malicious program wanted to get through the firewall, it'd just change its user agent from 'Terminator Virus' or whatever to 'Internet Explorer'.

**Comment by ghost** [comment]
My solution is to use Fiddler to chain to the corporate proxy. It can
rewrite user agents if need be.

**Comment by wilhelmberg** [comment]
> Is it possible to provide a newer build?

Yes, I'm planning on another deep dive into the proxy problem during the coming weeks.
But, at the moment I'm not able to specify an exact time frame.

**Comment by wilhelmberg** [comment]
@techieshark @voltagex 

Would be great, if you could give the latest proxy test version a spin and let me know how it goes.

Downloads and instructions https://github.com/mapbox/mapbox-studio/issues/860#issuecomment-112001682

**Comment by techieshark** [comment]
@BergWerkGIS Thanks for your work on this! I've been bike touring in New Zealand and don't have access to the system behind the proxy, but I'll talk to @Lilith-Palmer to see if anyone else can test this out over there.

**Comment by jahoover** [comment]
I pasted the information above to the command line and received the response below. I am operating behind a proxy (and use a pac) and cannot get mapbox studio to open. I have connected offline and then tried to connect but still no luck. Usually an etimedout error. 

Microsoft Windows [Version 6.1.7601]
Copyright (c) 2009 Microsoft Corporation.  All rights reserved.

C:\Users\mkjxh>SET HTTP_PROXY=http://10.150.1.109:8080

C:\Users\mkjxh>SET HTTPS_PROXY=http://10.150.1.109:8080

C:\Users\mkjxh>cd C:\Program Files\mapbox-studio

C:\Program Files\mapbox-studio>mapbox-studio.exe

C:\Program Files\mapbox-studio>[1456:1210/155026:INFO:renderer_main.cc(211)] Ren
derer process started
[10180:1210/155029:INFO:CONSOLE(0)] <Thu, 10 Dec 2015 20:50:29 GMT> Mapbox Studi
o @ http://localhost:3000/
[10288:1210/155050:INFO:renderer_main.cc(211)] Renderer process started
[7464:1210/155150:INFO:renderer_main.cc(211)] Renderer process started
[10180:1210/155151:INFO:CONSOLE(5)] "Uncaught ReferenceError: $ is not defined",
 source: https://assets-cdn.github.com/assets/frameworks-b7c9523d93e6cd0235a0a8b
7d1a691f7909d3834cd9c9caa7b1674ba508b644c.js (5)
[10180:1210/155151:INFO:CONSOLE(1)] "Uncaught ReferenceError: jQuery is not defi
ned", source: https://assets-cdn.github.com/assets/github-b42430c121e963c7370a99
d029c78f3c2fdc9b75b73adf817f07802e33988e81.js (1)

**Comment by wilhelmberg** [comment]
@jahoover proxy configuration via PAC is not supported and I don't think I can offer any help here as it seems your configuration lets some requests pass and others not: some assets are loaded but not the jquery module.

You might want to try the new Mapbox Studio that runs in your browser and doesn't need a local install:
https://www.mapbox.com/mapbox-studio/

**Comment by jahoover** [comment]
Wilhelm,
Thank you for your suggestion. My understanding however is that to add shape files and integrate with a base map for use in Tableau I need to use Mapbox Studio Classic. If that is not the case, do you have an article that describes providing this functionality via Mapbox studio?

jah

**Comment by Anuj-786** [comment]
Hey guys, I am working with cordova phonegap. I want to use mapbox in my app. when I added the cordova plugin for mapbox it gives me the error in console i.e Error: exec proxy not found for :: Mapbox :: show. have anyone idea about this. please help me.

**Comment by ghost** [comment]
@Anuj-786 That error is not related to this issue.
